bakery/postgresrootkeystore/rootkey.go

```diff
@@ -0,0 +1,347 @@
+	}
+	s.current[policy] = key
+}
+
+type store struct {
+	keys   *RootKeys
+	policy Policy
+	coll   *mgo.Collection
+}
+
+// Get implements bakery.RootKeyStore.Get.
+func (s *store) Get(ctx context.Context, id []byte) ([]byte, error) {
+	s.keys.mu.Lock()
+	defer s.keys.mu.Unlock()
+
+	key, err := s.keys.get(id)
+	if err != nil {
+		return nil, err
+	}
+	return key.rootKey, nil
+}
+
+// RootKey implements bakery.RootKeyStore.RootKey by
+// returning an existing key from the cache when compatible
+// with the current policy.
+func (s *store) RootKey(context.Context) ([]byte, []byte, error) {
+	if key := s.rootKeyFromCache(); key.isValid() {
+		return key.rootKey, key.id, nil
+	}
+	logger.Debugf("root key cache miss")
+	// Try to find a root key from the collection.
+	// It doesn't matter much if two concurrent mongo
+	// clients are doing this at the same time because
+	// we don't mind if there are more keys than necessary.
+	//
+	// Note that this query mirrors the logic found in
+	// store.rootKeyFromCache.
+	key, err := s.keys.findBestRootKey(s.policy)
+	if err != nil {
+		return nil, nil, errgo.Notef(err, "cannot query existing keys")
+	}
+	if !key.isValid() {
+		// No keys found anywhere, so let's create one.
+		var err error
+		key, err = s.generateKey()
+		if err != nil {
+			return nil, nil, errgo.Notef(err, "cannot generate key")
+		}
+		logger.Infof("new root key id %q; created %v; expires %v", key.id, key.created, key.expires)
+		if err := s.keys.insertKey(key); err != nil {
+			return nil, nil, errgo.Notef(err, "cannot create root key")
+		}
+	}
+	s.keys.mu.Lock()
+	defer s.keys.mu.Unlock()
+	s.keys.addCache(key.id, key)
+	s.keys.setCurrent(s.policy, key)
+	return key.rootKey, key.id, nil
+}
+
+// rootKeyFromCache returns a root key from the cached keys.
+// If no keys are found that are valid for s.policy, it returns
+// the zero key.
+func (s *store) rootKeyFromCache() rootKey {
+	s.keys.mu.Lock()
+	defer s.keys.mu.Unlock()
+	if k, ok := s.keys.current[s.policy]; ok && k.isValidWithPolicy(s.policy) {
+		return k
+	}
+
+	// Find the most recently created key that's consistent with the
+	// store policy.
+	var current rootKey
+	for _, k := range s.keys.cache {
+		if k.isValidWithPolicy(s.policy) && k.created.After(current.created) {
+			current = k
+		}
+	}
+	if current.isValid() {
+		s.keys.current[s.policy] = current
+		return current
+	}
+	return rootKey{}
+}
+
+func (s *store) generateKey() (rootKey, error) {
+	newKey, err := randomBytes(24)
+	if err != nil {
+		return rootKey{}, err
+	}
+	newId, err := randomBytes(16)
+	if err != nil {
+		return rootKey{}, err
+	}
+	now := timeNow()
+	return rootKey{
+		created: now,
+		expires: now.Add(s.policy.ExpiryDuration + s.policy.GenerateInterval),
+		// TODO return just newId when we know we can always
+		// use non-text macaroon ids.
+		id:      []byte(fmt.Sprintf("%x", newId)),
+		rootKey: newKey,
+	}, nil
+}
+
+func randomBytes(n int) ([]byte, error) {
```

Comment:
this could be moved to a common package as it is probably the same as in mgorootkeystore

Reply:
same for afterEq and beforeEq

Reply:
Too trivial to be worth it, I think. I'd like to factor out lots more code than that ideally :)